src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnectionInitiator.cs

```diff
@@ -0,0 +1,437 @@
+                                SR.EndpointNotFound, uri.AbsoluteUri), exception));
+                        }
+
+                        path = PipeUri.GetParentPath(path);
+                    }
+                }
+            }
+
+            if (string.IsNullOrEmpty(matchPipeName))
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                    new EndpointNotFoundException(SR.Format(SR.EndpointNotFound, uri.AbsoluteUri),
+                    new PipeException(SR.Format(SR.PipeEndpointNotFound, uri.AbsoluteUri))));
+            }
+
+            return matchPipeName;
+        }
+
+        private void PrepareConnect(Uri remoteUri, TimeSpan timeout, out string resolvedAddress, out BackoffTimeoutHelper backoffHelper)
+        {
+            PipeUri.Validate(remoteUri);
+            if (DiagnosticUtility.ShouldTraceInformation)
+            {
+                TraceUtility.TraceEvent(TraceEventType.Information, TraceCode.InitiatingNamedPipeConnection,
+                    SR.TraceCodeInitiatingNamedPipeConnection,
+                    new StringTraceRecord("Uri", remoteUri.ToString()), this, null);
+            }
+            resolvedAddress = GetPipeName(remoteUri);
+            const int backoffBufferMilliseconds = 150;
+            TimeSpan backoffTimeout;
+            if (timeout >= TimeSpan.FromMilliseconds(backoffBufferMilliseconds * 2))
+            {
+                backoffTimeout = TimeoutHelper.Add(timeout, TimeSpan.Zero - TimeSpan.FromMilliseconds(backoffBufferMilliseconds));
+            }
+            else
+            {
+                backoffTimeout = Ticks.ToTimeSpan((Ticks.FromMilliseconds(backoffBufferMilliseconds) / 2) + 1);
+            }
+
+            backoffHelper = new BackoffTimeoutHelper(backoffTimeout, TimeSpan.FromMinutes(5));
+        }
+
+        private async ValueTask<IConnection> TryConnectAsync(Uri remoteUri, string resolvedAddress, BackoffTimeoutHelper backoffHelper)
+        {
+            bool lastAttempt = backoffHelper.IsExpired();
+            // NamedPipeClientStream opens a pipe with the name "\\{serverName}\pipe\{pipeName}". As we only connect
+            // to local pipes, we pass "." as the serverName.
+            // PipeDirection means the pipe is opened READ/WRITE
+            // PipeOptions.Asynchronous sets the flag FILE_FLAG_OVERLAPPED and binds the handle to the threadpool IOCP
+            // TokenImpersonationLevel.Anonymous sets the SECURITY_QOS_PRESENT flag and disables impersonation 
+            // HandleInheritability.None sets the security attributes to null
+            NamedPipeClientStream namedPipeClient;
+            try
+            {
+                namedPipeClient = new NamedPipeClientStream(".", resolvedAddress, PipeDirection.InOut, PipeOptions.Asynchronous, TokenImpersonationLevel.Anonymous, HandleInheritability.None);
+                await namedPipeClient.ConnectAsync((int)backoffHelper.OriginalTimeout.TotalMilliseconds);
```

Comment:
Shouldn't this be the remaining time? Maybe I don't understand how timeouts work here. I would expect that the method immediately throws as soon as a timeout is exceeded.

Reply:
You're right that this is wrong. I looked at the implementation and it tries to connect to the named pipe, but that requires the service side to have called ConnectNamedPipe to have a pending connection. If that's not the case the connect fails and it calls WaitNamedPipe with the timeout. This synchronously blocks until ConnectNamedPipe is called by the service. If I pass zero, it uses the what was specified when the service created the pipe, which defaults to 50ms (and WCF doesn't set it to anything explicit). I'm going to pass 1 so that it doesn't hold a thread blocking waiting for the service to begin listening.